core/src/workflow/workflow_tasks/mod.rs

```diff
@@ -186,6 +198,67 @@ impl WorkflowTaskManager {
         }
     }
 
+    /// Returns number of currently cached workflows
+    pub fn cached_workflows(&self) -> usize {
+        self.workflow_machines.cached_workflows()
+    }
+
+    /// Resolves once there is either capacity in the cache, or there are no pending evictions.
+    /// Inversely: Waits while there are pending evictions and the cache is full.
+    /// Waiting while there are no pending evictions must be avoided because it would block forever,
+    /// since there is no way for the cache size to be reduced.
+    pub async fn wait_for_cache_capacity(&self) {
```

Comment:
Does this ever get called concurrently?
Is there a scenario where we have 2 waiters on cache capacity and we go over cap?

Reply:
That's why, inside the cache manager, there's a lock held while waiting for things to change. All waiters on cache capacity are effectively serialized (though I just realized I can get rid of the watch channel since it can just be a normal channel inside that mutex)